### PR TITLE
Suppression plages ouvertures

### LINKED ISSUE #2227: Supprimer les anciennes plages d'ouverture et indisponibilités après un délai d'un mois
## Actuellement

Nous gardons l'historique des plages d'ouverture et des indisponibilités.

Screenshot 2022-03-07 at 15-40-45 Vos plages d'ouvertures - RDV Solidarités

C'est très pratique pour pouvoir en créer une nouvelle par copie d'une ancienne.

## Objectif

Il y a parfois des informations sensibles dans les indisponibilités (et les plages d'ouvertures ?). Les garder indéfiniment n'est pas une bonne chose.

Mettre en place une suppression automatique des plages d'ouverture et des indisponibilités de plus de 1 mois.

### PR DESCRIPTION
**Pull request by NFontelle**
Close #2227

Mise en place de la suppression des plages d'ouvertures fermées depuis plus d'un mois
Mise en place des tests

AVANT LA REVUE
- [x] ~~Préparer des captures de l’interface avant et après~~
- [x] Nettoyer les commits pour faciliter la relecture
- [x] Supprimer les éventuels logs de test et le code mort

REVUE
- [x] Relecture du code
- [x] Test sur la review app / en local

### CODE REVIEW

File: spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+describe CronJob::DestroyOldPlageOuvertureJob do
+  it "Destroy exceptional po closed since 2 months" do
+    now = Time.zone.parse("20220405 10:00")
+    travel_to(now)
+    po_exceptionelle_closed_since_2_months = create(:plage_ouverture, first_day: now - 2.months, recurrence: nil)
```
Comment (francois-ferrandis) [inline spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb:7]: Je ne suis pas sûr de comprendre pourquoi un voyage dans le temps est nécessaire ici, n'aurait-on pas le même résultat avec ceci ?

```suggestion
    po_exceptionelle_closed_since_2_months = create(:plage_ouverture, first_day: 2.months.ago, recurrence: nil)
```

Ça s'applique aussi aux autres exemples. :wink:

**Review by francois-ferrandis — COMMENTED** [review]
Beau boulot ! J'ai juste une petite question sur les tests. :slightly_smiling_face:

File: spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb (hunk 1 above)
Comment (NFontelle) [reply spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb:7]: Pour cette PR Yannick m'a beaucoup aidé (seule je n'aurai pas pu faire ça) dans sa méthode de travail, (si j'ai bien compris sa démarche) il préfère avoir une date fixe de référence d'ou le "now" et le "travel_to"

File: app/jobs/cron_job.rb [hunk 2]
```diff
@@ -69,6 +69,17 @@ def perform
...
   end
 
+  class DestroyOldPlageOuvertureJob < CronJob
+    # At 01:00 every day
+    self.cron_expression = "0 1 * * *"
+
+    def perform
+      po_exceptionnelle_closed_since_1_month = PlageOuverture.where(recurrence: nil).where(first_day: ..1.month.ago)
+      po_reccurent_closed_since_1_month = PlageOuverture.where(recurrence_ends_at: ..1.month.ago)
+      po_exceptionnelle_closed_since_1_month.or(po_reccurent_closed_since_1_month).destroy_all
```
Comment (n-b) [inline app/jobs/cron_job.rb:79]: `PlageOuverture` inclut `WebhookDeliverable`: ça veut dire que quand on “destroy” une plage d’ouverture, ça va déclencher un appel de webhook.

C’est peut-être ce qu’on veut, mais dans le cas présent, ça va être assez violent à la première exécution quand il va falloir supprimer toutes les vieilles PO.

File: spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb (hunk 1 above)
Comment (yaf) [reply spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb:7]: Je trouve que voyager dans le temps permet d'être sur du test, alors que le « from_now » ou des trucs liés à la date courante peut avoir des effets de bord (et donc faire des flaky_test :)).

Comment (francois-ferrandis) [reply spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb:7]: La question est de savoir si l'on veut que ce test nous informe des edge cases en échouant parfois. :wink: En utilisant des dates relatives, on s'assure que le comportement de l'appli est le bon à plusieurs moment, pas à un seul moment. Le test est moins reproductible, mais il teste beaucoup plus de cas (autant qu'il y a d'exécution du test).

Par exemple, sur un autre projet j'ai eu un test qui échouait fin février les années bissextiles, ça m'a permis de corriger un bug qui aurait impacté le comportement de l'app. :exploding_head: 

Je concède que dans le cas présent, la suppression des vieux RDV n'est pas "mission-critical" donc on peut prioriser la stabilité des tests. :+1: 

La concision du code est aussi un autre aspect du sujet : en voyant ce code, je me suis dit qu'il y avait peut-être une raison pour laquelle on voyageait à cette date, autre que la reproductibilité.

Bien sûr, tout ça n'est pas bien important, c'était juste pour contribuer à construire une culture technique avec vous. :pray:

Comment (yaf) [reply spec/jobs/cron_job/destroy_old_plage_ouverture_job_spec.rb:7]: Merci d'avoir posé la question, ne serait-ce que pour parler de la culture commune.

File: app/jobs/cron_job.rb (hunk 2 above)
Comment (francois-ferrandis) [reply app/jobs/cron_job.rb:79]: Mais dîtes-moi, en fait la problématique se pose aussi pour #2421, n'est-ce pas ? :thinking:

Comment (n-b) [reply app/jobs/cron_job.rb:79]: ahoui, complètement! Rdv est lui aussi `WebhookDeliverable` !

Comment (yaf) [reply app/jobs/cron_job.rb:79]: La réponse : non, nous ne voulons pas mettre à jour via les webhooks :)

**Review by yaf — APPROVED** [review]
